### PR TITLE
Displaying timezone on answers page

### PR DESCRIPTION
**Pull request by Diego-Zulu**
## Preview:

Screen Shot 2020-09-08 at 21 54 43
Screen Shot 2020-09-02 at 23 37 10

## Description:

Adding timezone functionality to the newdle answer page as mentioned on #208 . The display timezone used by default is the one obtained from the user's browser. Dates are kept both in the format received by the server and in a timezone sensitive format to show the user

## Todo:

- Active date marked on calendar seems to ignore user timezone, need to look more into it

## Checks:

- All tests pass ✅ 
- All linting rules pass ✅

## Manual testing:

Environment setup:

```
make
source ./.venv/bin/activate
make flask-server
make react-server
```

- Newdle with same timezone as user's has it's dates displayed without changes ✅
- Newdle with option 17:00 - 17:30 on CEST displays as 12:00 - 12:30 on Sao Paulo time ✅
- Newdle displays date correctly with 12 hour difference (e.g. 10:00 August 31st on Japan should be 22:00 August 30th on Sao Paulo) ✅
- Same 12 hour difference check for busy times ✅
- Active date on calendar matches event date correctly when 12 hour difference ✅
- Active date on calendar matches event date correctly when 12 hour difference ✅
- "Accept all options where I'm available" toggle highlights only the aforementioned slots ✅

## Questions:

N/A

### CODE REVIEW

File: requirements.txt
Comment (Diego-Zulu) [inline requirements.txt:15]: I wanted to leave this change in for a discussion: Do we need to compile psycopg2 when installing dependencies? If not, we should start using `psycopg2-binary`

File: newdle/core/app.py
Comment (Diego-Zulu) [inline newdle/core/app.py:89]: SKIP_LOGIN was failing because `_configure_multipass` redirects were in effect, so had to add this to be able to use this option

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [inline newdle/client/src/util/date.js:20]: I belive busyTimes were not being considered for the answers page, so we wouldn't see them on the answers page even though react would try to renderize them. We should either: 
1 - Include them correctly on the display between a correct min and max hours (this approach)
2 - Don't show busy times that are outside min and max hours range

File: newdle/client/src/components/creation/timeslots/Timeline.js
Comment (Diego-Zulu) [inline newdle/client/src/components/creation/timeslots/Timeline.js:74]: I believe this `duration` argument was not being captured by the function so I removed it

File: newdle/auth.py
Comment (ThiefMaster) [inline newdle/auth.py:20]: `after_logout_url` would be a better name here

File: newdle/client/src/selectors.js
Comment (ThiefMaster) [inline newdle/client/src/selectors.js:18]: Since we never have a user timezone in the state, i would only return the guessed one.

We can add it back if that changes at some point in the future.

File: newdle/client/src/components/answer/AnswerPage.js
Comment (ThiefMaster) [inline newdle/client/src/components/answer/AnswerPage.js:215]: I think it'd be important to show the newdle's timezone as well, so the user knows that the event the newdle is about is taking place in timezone X, but the times are shown to them in timezone Y.

File: newdle/client/src/util/date.js
Comment (ThiefMaster) [reply newdle/client/src/util/date.js:20]: I don't think we care much about busy times outside the displayed timespan here. So I'd just hide those.

File: newdle/client/src/components/creation/timeslots/Timeline.js
Comment (ThiefMaster) [reply newdle/client/src/components/creation/timeslots/Timeline.js:74]: Indeed, thanks!

File: newdle/client/src/util/date.js
Comment (ThiefMaster) [inline newdle/client/src/util/date.js:30]: I don't think we can have any case with no `newdleTz` being set.

Actually, can `userTz` even be empty? I don't see anything about it in [the docs](https://momentjs.com/timezone/docs/#/using-timezones/guessing-user-timezone/).

So `userTz !== newdleTz` may be a better check here. But then I have the feeling that the logic here would simply not change anything, so maybe we don't need this check since we can always run the code here?

**Comment by ThiefMaster** [comment]
> Regarding busy times: I understand that (because I'm using the skip sign in option) I'm getting random busy times. I'm I correct to assume that the returned busy times' timezone is on the same timezone we provide as a parameter?

Yes. Here's an example with real data:

```
$ http -pb get 'https://newdle.blackhole.cern.ch/api/newdle/7kYbqYp9/participants/xxx/busy' 'date==2020-09-09' 'tz==UTC'
[
    [
        "08:30",
        "10:00"
    ],
    [
        "14:00",
        "15:00"
    ],
    [
        "22:00",
        "23:59"
    ]
]

$ http -pb get 'https://newdle.blackhole.cern.ch/api/newdle/7kYbqYp9/participants/xxx/busy' 'date==2020-09-09' 'tz==Europe/Zurich'
[
    [
        "10:30",
        "12:00"
    ],
    [
        "16:00",
        "17:00"
    ]
]
```

This particular user has the following busy times (in Europe/Zurich):

- Sep 9: 10:30-12:00, 16:00-17:00
- Sep 10: Full day (this is what causes the 22:00-23:59 entry for the 9th in ?UTC)

File: newdle/client/src/util/date.js
Comment (ThiefMaster) [reply newdle/client/src/util/date.js:20]: BTW, this indeed looks a bit weird (in master):

image

```
$ http -pb get 'https://newdle.blackhole.cern.ch/api/newdle/7kYbqYp9/participants/xprFxn2f/busy' 'date==2020-09-11' 'tz==Europe/Zurich'
[
    [
        "00:00",
        "00:00"
    ],
    [
        "14:00",
        "16:00"
    ]
]
```

However, I think this is actually a bug in the API. We should not return a zero-duration busy slot at all (this is coming from the all-day-busy on the day before).

File: newdle/client/src/components/answer/Calendar.js
Comment (ThiefMaster) [inline newdle/client/src/components/answer/Calendar.js:108]: `...ForTimezone` sounds a bit strange. In Indico we usually use "local" to indicate times in a local timezone, e.g. `startMomentLocal`. Or `startMomentUserTz` / `startMomentUser`? `...Local` sounds least strange to me though.

Anyway, naming discussions are not really important right now (and they tend to be harder than the actual technical discussions, because naming is a [hard problem](https://twitter.com/secretGeek/status/7269997868))

File: newdle/client/src/util/date.js
Comment (ThiefMaster) [reply newdle/client/src/util/date.js:20]: BTW, taking busy times into account here in any regard would probably be something for a separate issue/PR. Especially when it comes to timezone logic which is already tricky by itself, it does not make PRs easier to review and test if they include other, unrelated things as well. 

But nonetheless, thanks for realizing that this is a point which should probably be improved!

**Comment by ThiefMaster** [comment]
I was having another look at the free-busy timezone behavior and since the API returns it in the specified timezone,. you do not need to perform any client-side timezone conversions there - you just need to pass the `userTz` instead of the `newdleTz` when fetching fetching the free-busy information from the server. This means some of the new logic you added can actually be simplified a bit (and some of it maybe even removed again, such as the code added to `getBusySlotProps`). :)

**Comment by ThiefMaster** [comment]
FYI, I just opened #219 which correctly handles timezone information in the random free-busy data the API returns. I will merge this in a few minutes once CI passed, so once you rebased you should be able to test properly with the random data.

File: requirements.txt
Comment (Diego-Zulu) [reply requirements.txt:15]: Sounds good! I did not know psycopg2-binary was not recommended for prod environment. I'll remove the changes on this requirements.txt 👍

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [reply newdle/client/src/util/date.js:20]: Ok, I agree, I'll leave it out for now and create a separate issue

File: newdle/auth.py
Comment (Diego-Zulu) [reply newdle/auth.py:20]: Ok, I'll create a separate issue and PR, so I can move this code and the redirect problems when skipping login

File: requirements.txt
Comment (ThiefMaster) [reply requirements.txt:15]: It's mentioned in its [readme on pypi](https://pypi.org/project/psycopg2-binary/) ;)

> The binary package is a practical choice for development and testing but in production it is advised to use the package built from sources.

<details>
FWIW, they may be a bit overly cautious - in older psycopg2 versions binary wheels were the default so many people ended up using those implicitly in the normal psycopg2 package.
</details>

File: newdle/client/src/selectors.js
Comment (Diego-Zulu) [reply newdle/client/src/selectors.js:18]: Sounds good, I did it just to show how we could receive it in the future, but I agree "code speculation" is confusing. I'll leave only moment.tz.guess

File: newdle/client/src/components/answer/AnswerPage.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/AnswerPage.js:215]: Good idea, thanks!

File: newdle/auth.py
Comment (ThiefMaster) [reply newdle/auth.py:20]: No need for an issue, PR is enough for something as simple as this. :)

File: newdle/client/src/components/answer/Calendar.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/Calendar.js:108]: Ok, I'll revisit the naming. I think startTimeLocal should work 👍

**Comment by Diego-Zulu** [comment]
Ok, thanks for the comments and for the random provider free-busy data fix! I'll get to all of the changes mentioned above later today

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [reply newdle/client/src/util/date.js:30]: I forgot to address this comment. The problem here is that this same method is used during creation, so it won't receive any userTz or newdleTz. I'll revisit this to see if I can have a cleaner logic for both the creation and the answer pages

File: newdle/client/src/components/answer/AnswerPage.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/AnswerPage.js:218]: Added a simple "Newdle timezone" header, will update the PR's preview image ASAP

File: newdle/client/src/components/answer/Calendar.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/Calendar.js]: Immplemented startTimeLocal and startMomentLocal

File: newdle/core/app.py
Comment (Diego-Zulu) [reply newdle/core/app.py]: Moved this code to https://github.com/indico/newdle/pull/220

File: newdle/auth.py
Comment (Diego-Zulu) [reply newdle/auth.py]: Moved to https://github.com/indico/newdle/pull/220

File: newdle/auth.py
Comment (Diego-Zulu) [reply newdle/auth.py]: Done on https://github.com/indico/newdle/pull/220

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [reply newdle/client/src/util/date.js]: I noticed that I could just move the timezone for the outputted hours from the original `getHourSpan` method instead of modifying it

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [reply newdle/client/src/util/date.js]: Hey, I realized this discussion all steamed from me not understanding that busy slots below minHour will be assigned pos 0 and no height. I believed the minHour should be the minimum of the busy slots and options slots just because I didn't understand that I was rendering the busy slots wrongly. I now believe my timezone aware busySlots implementation is correct and they render correctly (pos and height wise), so I believe this discussion can be dropped

Comment (Diego-Zulu) [reply newdle/client/src/util/date.js]: Although I can open a new issue if we want to fix the red line that appears at the top when the busy slot is outside our display range

Comment (ThiefMaster) [reply newdle/client/src/util/date.js]: i think they are fine - to indicate that there is *something* earlier, without cluttering the vertical space.

(i'd just fix the 00:00-00:00 entries sent e.g. by the exchange backend, because those are useless since they come from the previous day)

File: newdle/client/src/components/answer/Option.js
Comment (ThiefMaster) [inline newdle/client/src/components/answer/Option.js:32]: `startTime` and `endTime` are no longer used so I would either:

- remove them from `propTypes` as well or
- keep using the original `startTime` and `endTime` props since AFAICT nothing changed inside this component besides the removal of the serialize+toMoment calls. BTW, the only reason these calls exist is to strip leading zeroes from hours before 10am, and that logic should probably be preserved since it's more human-friendly

File: newdle/client/src/components/answer/Calendar.js
Comment (ThiefMaster) [inline newdle/client/src/components/answer/Calendar.js:179]: I think you don't need any of the changes here if you make this change instead:

```patch
diff --git a/newdle/client/src/components/answer/AnswerPage.js b/newdle/client/src/components/answer/AnswerPage.js
index c25342d..87b7212 100644
--- a/newdle/client/src/components/answer/AnswerPage.js
+++ b/newdle/client/src/components/answer/AnswerPage.js
@@ -150,9 +150,9 @@ export default function AnswerPage() {

   useEffect(() => {
     if ((participantCode && !participantUnknown) || (!participantCode && user)) {
-      dispatch(fetchBusyTimesForAnswer(newdleCode, participantCode || null, dates, newdleTz));
+      dispatch(fetchBusyTimesForAnswer(newdleCode, participantCode || null, dates, userTz));
     }
-  }, [dates, newdleCode, participantCode, participantUnknown, user, newdleTz, dispatch]);
+  }, [dates, newdleCode, participantCode, participantUnknown, user, userTz, dispatch]);

   if (!newdle || (participantCode && !participant)) {
     return null;
```

(you probably need to still change some params or undo changes related to the new props you added )

File: newdle/client/src/util/date.js
Comment (Diego-Zulu) [reply newdle/client/src/util/date.js]: I think I fixed this 00:00-00:00 shortcoming, please check it out

File: newdle/client/src/components/answer/Option.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/Option.js]: You are right! I believed startTime and endTime were still being used as part of posting the availability to the backend, but then I noticed the slot has its own action already associated for that purpose. I deleted `startTimeLocal` and `endTimeLocal`, and went back to `startTime` and `endTime`

File: newdle/client/src/selectors.js
Comment (Diego-Zulu) [reply newdle/client/src/selectors.js]: I'll leave it as `export const getUserTimezone = () => moment.tz.guess();`. In the original issue it was discussed that it would be nice to allow the user to change this timezone, and I agree. So I want to leave it like this so a future maintainer doesn't need to worry about deprecating and removing function on `util/date.js`

File: newdle/client/src/components/answer/Calendar.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/Calendar.js]: You were totally right, so much code simplified, thanks for that!

File: newdle/client/src/selectors.js
Comment (ThiefMaster) [reply newdle/client/src/selectors.js]: good point! i wasn't thinking about that case but it makes sense to store the custom timezone in the state so keeping the selector is indeed a good idea :)

File: newdle/api.py
Comment (Diego-Zulu) [inline newdle/api.py:201]: Hey, let's discuss this piece of code. You see, when updating the active date code to correctly reflect the active date on the answer calendar when dealing with 12 hour diff timezones (japan - sao paulo for example), I noticed a problem here. My test case is:

- Newdle on Sept 4th 10AM Japan
- I use Sao Paulo, so I see displayed Sept 3rd 10PM

In this case, I end up asking my busy slots for Sept 3rd, and this check would fail because all slots were for Sept 4th. The solutions I thought about were:

1. This solution I wrote: Cast the slots timezone before asking for the date. Depending on the demand of this particular endpoint, this could be a costly operation
2. Ask for busy time on the timezone and date of the newdle, and then cast and aggregate them on the frontend as I did in a previous commit we discussed
3. Remove this check and call `_get_busy_times` even if there are no `participant.newdle.timeslots`

What do you think we should do?

Comment (ThiefMaster) [reply newdle/api.py:201]: I like your solution. It's simple, and timezone conversions are so cheap that they are perfectly fine. Every database query etc is likely more expensive than even a few timezone conversions.

Just one thing: We do not use camelCase in Python code - please use the more common `snake_case` in the code above.

File: newdle/client/src/components/answer/Option.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/Option.js:9]: Ok 👍 . I just fixed this, I decided to just serialize `startTime` at `getSlotProps` with single-digit hours there, as we aren't using it anywhere else

File: newdle/api.py
Comment (Diego-Zulu) [reply newdle/api.py]: Ok, sounds good! And I just fixed the camelCase on Python, I was pretty tired when I wrote that code and the camelCase totally passed me by. Thanks for the catch!

**Comment by Diego-Zulu** [comment]
Hey, I'm removing the Draft status form this PR. I have run over a couple examples myself and I all of them rendered correctly, unbeknownst to the timezone difference

Please feel free to continue commenting until you think things look good enough to merge! I'll continue addressing the comments ASAP

File: newdle/client/src/components/ParticipantTable.js
Comment (ThiefMaster) [inline newdle/client/src/components/ParticipantTable.js:118]: It looks good, but I wonder if it wouldn't make more sense to display the timezone in a single place instead of for every single timeslot (since it's always the same anyway).

Not a blocker for this PR though!

File: newdle/client/src/components/answer/AnswerPage.js
Comment (ThiefMaster) [inline newdle/client/src/components/answer/AnswerPage.js:218]: Should we display this in case it's the same as `newdleTz`? And I'm not sure if `Header` (especially when rendered as a h4 or any other hX element) is semantically correct - it's not really a header..

**Comment by ThiefMaster** [comment]
Thanks a lot, pretty much everything works great now! I only noticed one small but in the "accept all options when i'm available" feature. This only works you are in the same timezone as the newdle, as it uses newdle-tz timeslots with user-tz busy times.

Here's a fix that works for me - maybe you have a nicer idea, otherwise feel free to just commit that (ideally with the `--author` set to me ;))

```patch
diff --git a/newdle/client/src/answerSelectors.js b/newdle/client/src/answerSelectors.js
index 2877e14..d639194 100644
--- a/newdle/client/src/answerSelectors.js
+++ b/newdle/client/src/answerSelectors.js
@@ -32,6 +32,12 @@ export const getLocalNewdleTimeslots = createSelector(
       )
     )
 );
+/** A mapping from user-tz timeslots to newdle-tz timeslots */
+const getLocalNewdleTimeslotsMap = createSelector(
+  getNewdleTimeslots,
+  getLocalNewdleTimeslots,
+  (timeslots, localTimeslots) => _.zipObject(localTimeslots, timeslots)
+);
 export const getCalendarDates = createSelector(
   getLocalNewdleTimeslots,
   timeslots =>
@@ -81,17 +87,17 @@ export const isAllAvailableSelectedExplicitly = state => state.answer.allAvailab
 /** All time slots during which the person is free */
 const getFreeTimeslots = createSelector(
   getFlatBusyTimes,
-  getNewdleTimeslots,
+  getLocalNewdleTimeslotsMap,
   getNewdleDuration,
-  (busyTimes, timeslots, duration) => {
+  (busyTimes, timeslotMap, duration) => {
     busyTimes = busyTimes.map(pair => pair.map(t => toMoment(t, moment.HTML5_FMT.DATETIME_LOCAL)));
-    timeslots = timeslots.map(t => [
+    const localTimeslots = Object.keys(timeslotMap).map(t => [
       toMoment(t, moment.HTML5_FMT.DATETIME_LOCAL),
       toMoment(t, moment.HTML5_FMT.DATETIME_LOCAL).add(duration, 'm'),
     ]);
-    return timeslots
+    return localTimeslots
       .filter(ts => !busyTimes.some(bt => overlaps(ts, bt)))
-      .map(([start]) => serializeDate(start, moment.HTML5_FMT.DATETIME_LOCAL))
+      .map(([start]) => timeslotMap[serializeDate(start, moment.HTML5_FMT.DATETIME_LOCAL)])
       .sort();
   }
 );
```

File: newdle/client/src/components/ParticipantTable.js
Comment (Diego-Zulu) [reply newdle/client/src/components/ParticipantTable.js:118]: I adds a lot of clutter in the case you have many slots, but I believe that it also helps people not loosing track about the original timestamp. But I do also see a use case for rendering it only once, specially because in the future we may decide to display this time in the ParticipantTable in the user's local timezone and we could use that same place to indicate the difference in timezones

For now, I noticed that I can display an abbreviation for a given timezone, so I'll go with that to reduce clutter

Comment (ThiefMaster) [reply newdle/client/src/components/ParticipantTable.js:118]: What kind of abbreviation do you mean? If it's `UTC+xx` then it may be confusing for people, so I'd rather keep the repeated long `Europe/Zurich` etc

File: newdle/client/src/components/answer/AnswerPage.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/AnswerPage.js:218]: 1) Yes, I believe we should still display both of them, to prevent confusion from people
2) I agree the semantical meaning is not correct, so I'll change it to something that semantically makes more sense

File: newdle/client/src/components/ParticipantTable.js
Comment (Diego-Zulu) [reply newdle/client/src/components/ParticipantTable.js:118]: Okay, I agree, abbreviation are not as cool as I thought 😄 . Sao Paulo timezone's abbreviation is just `(-03)`, so I agree that leaving the long repeated timezone is better

Comment (ThiefMaster) [reply newdle/client/src/components/ParticipantTable.js:118]: yeah, let's keep it as-is then :)

File: newdle/client/src/components/answer/AnswerPage.js
Comment (Diego-Zulu) [reply newdle/client/src/components/answer/AnswerPage.js]: Done, take a look at it now

**Review by ThiefMaster — APPROVED** [review]
Great PR, thanks a lot!

**Pull request merged**
